lib/task_005.rb

```diff
@@ -48,52 +48,76 @@
 module Task005
   module_function
 
-  def parse_input(input_data)
-    input_data.split("\n").map { |line| parse_line(line) }
-  end
+  class Point
+    attr_reader :x, :y
 
-  def parse_line(line)
-    point1, point2 = line.split(' -> ')
+    def initialize(x, y)
+      @x = x
+      @y = y
+    end
 
-    {
-      point1: parse_point(point1),
-      point2: parse_point(point2),
-    }
+    def self.build_from_str(point)
+      x, y = point.split(',').map(&:to_i)
+      Point.new(x, y)
+    end
   end
 
-  def parse_point(point)
-    point.split(',').map { |value| value.to_i }
-  end
+  class Line
+    attr_reader :point1, :point2
 
-  def mark_lines(lines)
-    diagram = {}
-    lines.each { |line| mark_line(line, diagram) }
-    diagram
+    def initialize(point1, point2)
+      @point1 = point1
+      @point2 = point2
+    end
+
+    def self.build_from_hash_line(line)
+      Line.new(
+        Point.build_from_str(line.split(' -> ')[0]),
+        Point.build_from_str(line.split(' -> ')[1]),
+      )
+    end
   end
 
-  def mark_line(line, diagram)
-    if line[:point1][0] == line[:point2][0] || line[:point1][1] == line[:point2][1]
-      x1, x2 = [line[:point1][0], line[:point2][0]].sort
-      y1, y2 = [line[:point1][1], line[:point2][1]].sort
-      (x1..x2).each do |x|
-        (y1..y2).each do |y|
-          diagram.key?([x, y]) ? diagram[[x, y]] += 1 : diagram.store([x, y], 1)
+  class Diagram
+    attr_accessor :diagram
+
+    def initialize(diagram = {})
+      @diagram = diagram
+    end
+
+    def mark_lines(lines)
+      lines.each { |line| mark_line(line, diagram) }
+      diagram
+    end
+
+    def mark_line(line, diagram)
+      if line.point1.x == line.point2.x || line.point1.y == line.point2.y
+        x1, x2 = [line.point1.x, line.point2.x].sort
+        y1, y2 = [line.point1.y, line.point2.y].sort
+        (x1..x2).each do |x|
+          (y1..y2).each do |y|
+            diagram.key?([x, y]) ? diagram[[x, y]] += 1 : diagram.store([x, y], 1)
+          end
```

Comment:
This code is part of Diagram class.

And the moment Diagram has some knowledge how the lines work:
- it knows how to check if the line is vertical or horizontal,
- it knows how to find all points that lying on the line.

I think such knowledge could be hidden in `line` object. So, for example Diagram could only do:

Reply:
Here is what I mean:

```suggestion
      if line.horizontal_or_vertical?
        line.points_on_line.each do |point|
          @diagram.key?([x, y]) ? diagram[[x, y]] += 1 : diagram.store([x, y], 1)
        end
```

Reply:
correct, done :)